pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,300 @@
+	PriorityLevelConfigurationConditionUnknown PriorityLevelConfigurationConditionStatus = "Unknown"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchema defines the schema of a group of flows. Note that a flow makes up of a set of inbound API requests with
+// similar attributes and is identified by a pair of strings: the name of the FlowSchema and a "flow distinguisher".
+type FlowSchema struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+	// Specification of the desired behavior of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec FlowSchemaSpec
+	// Current status of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status FlowSchemaStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchemaList is a list of FlowSchema objects.
+type FlowSchemaList struct {
+	metav1.TypeMeta
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+
+	// Items is a list of flow-schemas.
+	Items []FlowSchema
+}
+
+// FlowSchemaSpec describes how the flow-schema's specification looks like.
+type FlowSchemaSpec struct {
+	// PriorityLevelConfigurationRef can reference a PriorityLevelConfiguration in the cluster. If the reference cannot
+	// be resolved, the flow-schema will ignored and marked as invalid in its status.
+	PriorityLevelConfigurationRef PriorityLevelConfigurationRef
+	// MatchingPriority controls the ordering the flow-schemas in the cluster. Its value must be greater than zero. The
+	// primary sort is based on MatchingPriority, ordered highest to lowest (20 before 10).
+	MatchingPriority int32
+	// Distinguisher defines how to classify requests into certain flows. If not empty, the flow's identifier is computed
+	// according to specification of the distinguisher.
+	// +optional
+	Distinguisher *FlowDistinguisher
+	// Rules holds all the RuleWithUser for this FlowSchema.
+	Rules []RuleWithUser
+}
+
+// FlowDistinguisher specifies the configuration of the distinguisher.
+type FlowDistinguisher struct {
+	// Source defines the source attribute we adopt to compute the identifier for the flow.
+	Source FlowDistinguisherSource
+}
+
+// FlowDistinguisherSource defines the input source of distinguishing flows.
+type FlowDistinguisherSource string
+
+// PriorityLevelConfigurationRef contains information that points to the request-priority being used.
+type PriorityLevelConfigurationRef struct {
+	// Name is the name of resource being referenced
+	Name string
+}
+
+// Rule is a trio of User, Verb and RequestScope.
```

Comment:
It looks like the idea here is to move closer to the way things are done in RBAC.  If so, why not import and use RBAC's `PolicyRule`?

I must admit I do not understand the design of RBAC's `Subject`.  Service accounts are just one of two kinds of users; why are they called out separately?  And what useful information ever appears in `Subject.APIGroup` or `Subject.Namespace` that is not already contained in or implied by `Subject.Name`?

Reply:
> It looks like the idea here is to move closer to the way things are done in RBAC. If so, why not import and use RBAC's PolicyRule?

tho their model looks really similar, but they belong to completely different api group for different usage. i don't see advantage from importing rbac except for minimizing the source code.

Reply:
I think the meaning of RBAC's `PolicyRule` is exactly what we are trying to use here, right?  If so, then I think it would be appropriate to import the type.